sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,140 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_hmac.h"
+
+#include "hmac_regs.h"  // Generated.
+
+// A hack to appease the powers that be in hmac_regs.h
+#define HMAC0_BASE_ADDR 0x0
+
+#define HMAC_FIFO_MAX 16
+
+static uint32_t get_status(const dif_hmac_t *hmac) {
+  return mmio_region_read32(hmac->base_addr, HMAC_STATUS(0));
+}
+
+static uint32_t fifo_depth(const dif_hmac_t *hmac) {
+  return (get_status(hmac) >> HMAC_STATUS_FIFO_DEPTH_OFFSET) &
+         HMAC_STATUS_FIFO_DEPTH_MASK;
+}
+
+static uint32_t fifo_avail(const dif_hmac_t *hmac) {
+  return HMAC_FIFO_MAX - fifo_depth(hmac);
+}
+
+void dif_hmac_init(const dif_hmac_config_t *config, dif_hmac_t *hmac_out) {
+  uint32_t device_config = 0;
+
+  // Clear the config, stopping the SHA engine.
+  mmio_region_write32(config->base_addr, HMAC_CFG(0), device_config);
+
+  // Clear INTER_STATE.
+  mmio_region_write32(config->base_addr, HMAC_INTR_STATE(0),
+                      1 << HMAC_INTR_ENABLE_HMAC_DONE |
+                          1 << HMAC_INTR_STATE_FIFO_FULL |
+                          1 << HMAC_INTR_STATE_HMAC_ERR);
+
+  // Set the byte-order of the input message.
+  if (config->endian_swap) {
+    device_config |= 1 << HMAC_CFG_ENDIAN_SWAP;
+  }
+  // Set the byte-order of the digest.
+  if (config->digest_swap) {
+    device_config |= 1 << HMAC_CFG_DIGEST_SWAP;
+  }
+
+  // Set HMAC engine to HMAC or SHA256-only mode.
+  if (config->mode == DIF_HMAC_OP_HMAC) {
+    device_config |= 1 << HMAC_CFG_HMAC_EN;
+
+    // Set the key making no assumptions about register locations.
```

Comment:
Since the `HMAC_KEY#_REG_OFFSET` values are declared individually and reflect the offsets in the documentation I was taking the safest approach by not making any assumptions about the register layout. If this isn't a realistic concern I'll happily make this a loop.